app/shared/session/useLoginErrors.ts

```diff
@@ -0,0 +1,57 @@
+import { useEffect, useState } from "react";
+import { useSearchParams } from "react-router-dom";
+import { sentryException } from "../sentry";
+
+export const AUTH_PROVIDERS = {
+  LOGIN_DEV: "login_dev",
+  LOGIN_GITHUB: "login_github",
+  LOGIN_GOOGLE: "login_google",
+};
+
+export const LOGIN_ERROR_MESSAGES = {
+  [AUTH_PROVIDERS.LOGIN_DEV]: "There has been an issue logging you in with dev",
+  [AUTH_PROVIDERS.LOGIN_GITHUB]:
+    "There has been an issue logging you in with Github",
+  [AUTH_PROVIDERS.LOGIN_GOOGLE]:
+    "There has been an issue logging you in with Google",
+};
+
+export const useLoginErrors = (): string => {
+  const [searchParams, setSearchParams] = useSearchParams();
+  const [messageToReturn, setMessageToReturn] = useState("");
+  const error = searchParams.get("error");
+  const message = searchParams.get("message");
+
+  useEffect(() => {
+    const hasMessageToShow =
+      error !== null && Boolean(message) && typeof message === "string";
+    if (hasMessageToShow) {
+      sentryException({
+        message: message,
+      });
+      setMessageToReturn(message);
+
+      setSearchParams({});
+      return;
+    }
+
+    switch (error) {
+      case AUTH_PROVIDERS.LOGIN_DEV:
+        setMessageToReturn(LOGIN_ERROR_MESSAGES[AUTH_PROVIDERS.LOGIN_DEV]);
+        break;
+      case AUTH_PROVIDERS.LOGIN_GITHUB:
+        setMessageToReturn(LOGIN_ERROR_MESSAGES[AUTH_PROVIDERS.LOGIN_GITHUB]);
+        break;
+      case AUTH_PROVIDERS.LOGIN_GOOGLE:
+        setMessageToReturn(LOGIN_ERROR_MESSAGES[AUTH_PROVIDERS.LOGIN_GOOGLE]);
+        break;
+
+      default:
+        break;
+    }
+
+    setSearchParams({});
```

Comment:
setting state from effect is a typical way to get an infinite loop, its like a generally bad thing to have, even if it works for some reason here, its kindof dangerous, let me know if you need help finding a better way to handle it, don't have a solution out of top of my head

Reply:
I was just discussing today this problem its a nasty problem with useEffect and state, but we have to deal with it somehow, one way is to use useRef so that you have access to that value without having a dependency on the ref and setting ref can be done directly inside the component, its  kind ugly as well but the best I know so far:

something like this:

```ts
  const [searchParams, setSearchParams] = useSearchParams();
 const searchParamsRef = useRef()
searchParamsRef.current = searchParams

useEffect(() => {
  const error = searchParams.current.get("error");
  const message = searchParams.current.get("message");
}, [setSearchParams])

Reply:
I can tell you exactly why this does not rerender and its because the search params in the URL are not reactive, they don't work like state so it Wong cause a run of the use effect again 

I was gonna either keep it because i know it doesn't re render , i worked with URLs a lot at remote or make the use effect only run on start since that's all we cafe about

Reply:
ignore me then